2983: Add expectation types for error handling

kduske opened the pull request:
@ericwa I have implemented a few result types. These are not finished and some features do not yet work due to missing features from `nonstd::variant`, but I'd like to get some feedback anyway. This PR adds the result types and some tests, and I have used the result types in `Brush::buildGeometry` instead of throwing exceptions. clang now warns about every place where this function is used without error handling.

kduske commented:
Update: I have added the abseil library to get a working variant type on macOS -- nonstd is pretty broken, and now I don't trust any of their stuff for more advanced use cases, so I'd like to replace them all with something else. Not sure whether it's going to be abseil -- there are other options as well, such as boost or one of these hacks: https://stackoverflow.com/questions/52310835/xcode-10-call-to-unavailable-function-stdvisit/53868971#53868971

Anyway, it looks like the result types will work for returning by value and by reference, with optional values, or for void functions. I'm still debating whether we should have special handling for returning pointers - currently these would be returned by value, and return null would be a valid value and not an error. I'm leaning towards leaving that as is, because if you want to disallow null as a valid result value, you can use references to communicate this.

There's still a lot of cases to test -- combinations of const vs. non-const result, const vs non-const value / reference in the result, taking the value by const lvalue, lvalue, rvalue reference, etc. Not sure how, but I'd like to define a proper test strategy to cover all corner cases.

kduske commented:
Update: I have added more tests, hopefully these now cover the most common cases when working with these types. I have also decided to call these types `expected` instead of `result` because I found the latter to be too generic. If you have a better name, let me know.

ericwa commented:
I prefer result type, it's a well known name for these e.g. https://en.wikipedia.org/wiki/Result_type

ericwa commented on common/src/Model/Brush.cpp:
```diff
@@ -1289,11 +1296,13 @@ namespace TrenchBroom {
+                return kdl::expected<void, GeometryException>(GeometryException("Brush is empty"));
             } else  if (!addFacesToGeometry.brushValid()) {
-                throw GeometryException("Brush is invalid");
+                return kdl::expected<void, GeometryException>(GeometryException("Brush is invalid"));
```
I think the constructor should be deleted and you have to either use a `::success` or `::failure` static method, so it's absolutely clear which type is being constructed. This is how they work in Rust, Swift, etc.

I can already that you need to read the template arguments carefully to figure out which of these constructing a failure and success.

kduske replied:
See above.

ericwa commented on lib/kdl/include/kdl/expected.h:
```diff
@@ -0,0 +1,627 @@
+         * @return the result of applying the given visitor
+         */
+        template <typename Visitor>
+        friend auto visit(Visitor&& visitor, const expected& expected) {
```
Is it just a matter of style that this is a free function?
If it is, I prefer member functions - it gives you more information when reading calling code, but I'm ok with either.

kduske replied:
A matter of style and consistency with `std::visit`. I'd be ok with a member function, but that deviates from the style of the STL. Personally, I also tend to use more free functions now.

kduske replied:
Having thought about it more, I agree now since it is more consistent with `map` being a member function as well. Maybe we should rename the function, too, to distinguish it from `std::visit`. Any ideas about an alternative name?

ericwa replied:
Not sure. I think a member function `result.visit(..)` would be OK, it's not a error to reuse the name is it?

kduske replied:
One reason for a free function is that it's syntactically nicer to operator on an rvalue this way. Consider:

```
auto result = ...;
visit(overload{...}, std::move(result));
```

vs.

```
auto result = ...;
std::move(result).visit(overload{...});
```

I'll keep it a member function for now, but we might want to revert it if this becomes awkward to use.

ericwa commented on lib/kdl/include/kdl/expected.h:
```diff
@@ -0,0 +1,627 @@
+        }
+    };
+    
+    /**
```
I followed along up to this point, but I don't really understand why we need any of the further types.
Also it means they have to reimplement all of kdl::expected, port bugfixes between all the copied code, etc.

If you want to return a `int&` why not just make a type `kdl::expected<std::reference_wrapper<int>, ...>` in the calling code?

kduske replied:
Because this is easier to read and understand (IMO) for users of this class. It's true that it adds a burden of maintenance to the library, but I think it makes using these types far easier, since `result<int&, This, That>` expresses intent much better than `result<std::reference_wrapper<int>, This, That>`.

Same goes for `result<void, This, That>` vs. `result<success_tag, This, That>`, but I am not sure about `result<opt<int>, This, That>` vs. `result<std::optional<int>, This, That>`. I guess in the latter case, using `std::optional` instead of `opt` is clearer, however I like that a visitor of an optional result value can take a `int` directly without having to care whether the value actually exists.

ericwa replied:
Ah I see the goal of the `opt<>` one now. Although it feels like it could be implemented as a special `visit()` override on the basic `result<T>`, rather than a whole new class - it could even have a different name like `visit_and_unwrap_optional()`.

I'll need to actually use these a bit to comment more.. There are a lot of ways `visit()` could work. I suspect for `result<opt<int>` we'll want `visit()` to run even when the result contains an empty optional, and for `result<void` we'll want visit to run when it's in the "success" (empty) state.  e.g. `visit()` could take two lambdas instead of just one, a success and a failure lambda.

> since result<int&, This, That> expresses intent much better than result<std::reference_wrapper<int>, This, That>.

It's syntax sugar, not sure about expressing intent better. Hopefully we don't return references enough to need syntax sugar for this, since it's dangerous, but I guess it is a nicer syntax.

ericwa commented on lib/kdl/include/kdl/expected.h:
```diff
@@ -0,0 +1,627 @@
+         * Indicates whether this expectation contains a value.
+         */
+        bool success() const {
+            return absl::holds_alternative<value_type>(m_value);
```
Note, from https://en.cppreference.com/w/cpp/utility/variant/holds_alternative , "The call is ill-formed if T does not appear exactly once in Types..."

This could be written as `m_value.index()` == 0 to not have that restriction, but I guess other code does.

kduske replied:
You mean for the case when the value type is also used as an error type? Good point.

ericwa replied:
Yeah

kduske replied:
done

ericwa commented:
I think this is good, just a few details I'm not sure about.

Having 4 implementations is something we should really try to avoid. I think we can make it work with just the first one.

- for the `int&` case, user code can just use `kdl::expected<std::reference_wrapper<int>, ...>`.

- For the success type = `void` case, why don't we provide a `struct success_tag {};` instead? user code can just use `kdl::expected<kdl::success_tag, ...>`.

- I don't understand the point of the `opt<>` specialization, what's the idea there?

Most of the languages with Result in their standard lib just have 1 Error case, and then, I guess, you would further pattern match into that. But since the `visit` syntax is more heavy-weight in C++, I guess we want to avoid that.

kduske commented on common/src/Model/Brush.cpp:
```diff
@@ -1273,14 +1274,20 @@ namespace TrenchBroom {
+            }
+            
             nodePhysicalBoundsDidChange(oldBounds);
+            return kdl::expected<void, GeometryException>();
```
That's fine, I agree. Done.

kduske commented:
Thank you for your comments! I have applied some of your feedback directly (renamed to result, added the static factories). For some of your points, I have added some answers and am still undecided (see inline comments).

ericwa commented:
I think a rebase might have gone wrong, the Brush.cpp changes are gone (where throwing exceptions was changed to returning kdl::result types?) Or did you remove that intentionally?

ericwa commented:
> I'm still debating whether we should have special handling for returning pointers - currently these would be returned by value, and return null would be a valid value and not an error. I'm leaning towards leaving that as is, because if you want to disallow null as a valid result value, you can use references to communicate this.

Agreed, no special handling is needed.

ericwa commented:
I think this is good to merge once the build's fixed, I want to try this in my logging branch. We'll probably need to fine tune it a bit once we get more experience using it.

ericwa commented on lib/kdl/include/kdl/result.h:
```diff
@@ -0,0 +1,669 @@
+    public:
+        template <typename V, typename std::enable_if<std::is_convertible_v<V, Value>>::type* = nullptr>
+        static result success(V&& v) {
+            return result(std::forward<V>(v));
```
One more thing, how do you know this is going to construct the success case? (e.g. assume all the variant's types have default constructors.) Does it try them in order?

`result(value_type(std::forward<V>(v)));`

would eliminate any doubt I think. We need to also say in the docs that you can't use reuse your Value type as an error type.

kduske replied:
Using the same type more than once in the argument list of `result` yields a compile error when you try to construct it. I'll try to add more SFINAE type checking to give bettter error messages tho.

ericwa replied:
`success()` should pass `std::in_place_index_t<0>()` as the first argument to the variant constructor to force it to actually create the success case.

Making `error()` fully safe is trickier, I was just playing around on godbolt and got this:

```
#include <variant>
#include <string>
#include <type_traits>
#include <cassert>

template<class Value, class... Errors>
struct result {

    using VariantType = std::variant<Value, Errors...>;

    template <typename V>
    static VariantType make_value(V&& v) {
        // force creation of index 0 (Value)
        return VariantType(std::in_place_index_t<0>(), v);
    }

    template <typename V>
    static VariantType make_error(V&& v) {
        // check that V is one of Errors
        static_assert(std::disjunction_v<std::is_same<V, Errors>...>);

        return VariantType(std::in_place_type_t<V>(), v);
    }
};

int main() {

    // implicit conversion from float to int   
    auto v1 = result<int, float, std::string>::make_value(1.2f);
    assert(v1.index() == 0);

    auto v2 = result<int, float, std::string>::make_error(1.3f);
    assert(v2.index() == 1);

    auto v3 = result<int, float, std::string>::make_error(std::string("abc"));
    assert(v3.index() == 2);

    // compile error: make_error() wants float or std::string
    //auto v4 = result<int, float, std::string>::make_error(1);

    return 0;
}
```

kduske replied:
Ok, I have added similar code to our factory methods.

kduske replied:
Nice, I didn't know about `std::disjunction` yet -- I was looking for something like that the other day ;-)

kduske commented:
> I think a rebase might have gone wrong, the Brush.cpp changes are gone (where throwing exceptions was changed to returning kdl::result types?) Or did you remove that intentionally?

That was intentional. I also dropped the other commit that makes `rebuildBrushGeometry` private, it doesn't really belong into this branch.

kduske commented:
I have decided to revert the changes to `visit` and make it a free function, along with `map`. We will often want to treat result as rvalues (to move the contained value), and having those functions as members is very awkward.

Some remaining questions:
- Should we rename them to `visit_result` and `map_result` to distinguish them from std functions?
- The parameter order of `map` is a bit counter intuitive now (function first, then the value that's being mapped). But I'd like to keep it in line with `visit`, and I'd like to keep `visit` in line with the parameter order of `std::visit`...
- Should we use `absl` now? If we do, I'd completely replace nonstd with absl in a separate PR. For what it's worth, absl uses the std implementations if they are available.

kduske commented:
Regarding abseil: It looks like that brings a lot of dependencies into our build, so I'd rather look for an alternative. Maybe we should just do this: https://github.com/youknowone/xcode-libcxx

ericwa commented:
> Should we rename them to visit_result and map_result to distinguish them from std functions?

Agreed

> The parameter order of map is a bit counter intuitive now (function first, then the value that's being mapped). But I'd like to keep it in line with visit, and I'd like to keep visit in line with the parameter order of std::visit...

Sounds fine

>  Maybe we should just do this: https://github.com/youknowone/xcode-libcxx 
> It looks like that brings a lot of dependencies into our build

What kind of dependencies, just a lot of abseil headers?

I don't really have a preference of what we use.

BTW, I have 2 branches in progress (logging/prefs, soft bounds) that have nonstd::optioanl usage, but it's easy to swap those out if we switch to something else.

kduske commented:
I have removed absl and have added backported headers for any, optional and variant. I'll remove nonstd entirely in a separate PR.

Furthermore I have renamed visit and map as proposed.

kduske commented:
If the backported headers don‘t work, we can keep using nonstd for optional and any and this for variant: https://github.com/mpark/variant

kduske commented:
Thanks for your helpful comments, I think the code is much better now!
